etna/datasets/hierarchical_structure.py

```diff
@@ -0,0 +1,187 @@
+from collections import defaultdict
+from itertools import chain
+from queue import Queue
+from typing import Dict
+from typing import List
+from typing import Optional
+from typing import Set
+from typing import Union
+
+import scipy
+from scipy.sparse import lil_matrix
+
+from etna.core import BaseMixin
+
+
+class HierarchicalStructure(BaseMixin):
+    """Represents hierarchical structure for provided hierarchical tree."""
+
+    def __init__(self, level_structure: Dict[str, List[str]], level_names: Optional[List[str]] = None):
+        """Init HierarchicalStructure.
+
+        Parameters
+        ----------
+        level_structure:
+            Adjacency list describing the structure of the hierarchy tree (i.e. {"total":["X", "Y"], "X":["a", "b"], "Y":["c", "d"]}).
+        level_names:
+            Names of levels in the hierarchy in the order from top to bottom (i.e. ["total", "category", "product"]).
+            If None is passed, level names are generated automatically with structure "level_<level_index>".
+        """
+        self._hierarchy_root: Union[str, None] = None
+        self._hierarchy_interm_nodes: Set[str] = set()
+        self._hierarchy_leaves: Set[str] = set()
+
+        self.level_structure: Dict[str, List[str]] = level_structure
```

Comment:
We do't need to provide typings for the parameters of the constructor as the already have typings